text/0000-code-literals.md

```diff
@@ -0,0 +1,268 @@
+        }
+        ```
+
+        If the indentation is changed, such as by removing the
+        conditional, then `rustfmt` must re-format the code like so:
+
+        ```rust
+        do_something_with(
+            "
+                a nicely
+                indented code
+                string
+                "
+        );
+        ```
+
+        To do otherwise would be to change thange the value of
+        the string literal.
+    
+    2.  Normal string literals with backslash escaping, eg.
+        ```rust
+        "
+        this way\
+        whitespace at\
+        the beginning\
+        of lines can\
+        be ignored\
+        "
+        ```
+
+        This approach still suffers from the need to escape special
+        characters. The backslashes at the end of every line are
+        tedious to write, and are problematic if whitespace is
+        meaningful within the code. For example, if python code
+        was being embedded, then the indentation would be lost.
+        Finally, although `rustfmt` could in principle reformat
+        these strings, in practice doing so in a reasonable way
+        is complicated and so this has never been enabled by default.
+    
+    3.  Raw string literals, eg. `r#"I can use "s!"#`
+
+        This solves the problem of special characters, but suffers
+        from the same inability to be reformatted, and the trick
+        of using an `\` at the end of each line cannot be applied
+        because escape characters are not recognised.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In addition to string literals and raw string literals, a third type
+of string literal exists: code string literals.
+
+```rust
+    let code = ```
+        This is a code string literal
+
+        I can use special characters like "" and \ freely.
+
+            Indentation is preserved *relative* to the indentation level
+            of the first line.
+
+    It is an error for a line to have "negative" indentation (ie. be
+    indented less than the indentation of the opening backticks) unless
+    the line is empty.
+        ```;
+```
+
+`rustfmt` will automatically adjust the indentation of the code string
+literal as a whole to match the surrounding context, but will never
+change the relative indentation within such a literal.
+
+Anything directly after the opening backticks is not considered
+part of the string literal. It may be used as a language hint or
+processed by macros (similar to the treatment of doc comments).
+
+```rust
+let sql = ```sql
+    SELECT * FROM table;
+    ```;
+```
+
+Similar to raw string literals, there is no way to escape characters
+within a code string literal. It is expected that procedural macros
+would build upon code string literals to add support for such
+functionality as required.
+
+If it is necessary to include triple backticks within a code string
+literal, more than three backticks may be used to enclose the
```

Comment:
Hmm, I'm torn here.  Using the same thing as in doccomments makes sense, but at the same time when we already have "use more `#`s to escape more" I don't feel amazing about *also* having a "use more ``` ` ```s to escape more" construct.